commitizen-tools/commitizen
# feat(tags): adds `legacy_tag_formats` and `ignored_tag_formats` settings

## Description

This PR add 2 new settings to handle tag formats;
- `legacy_tag_formats` allowing tag format to change over time while retaining the ability to work with legacy version
- `ignored_tag_formats` to specify expected tag that should not raise `InvalidVersion`.

So projects which change tag formats and have expected but non-version tags can have these settings:

```toml
tag_format = "v${version}"
legacy_tag_formats = [
	"${version}",
	"project-prefix-${version}",
]
ignored_tag_formats = [
	"stable",
	"other-project-*",
]
```

By doing so, I had to start grouping tag handling and introduce a `tags.TagRules` object which is responsible for handling all tags rules, initially spread over `tag_format`, `version_scheme`s and `version_provider`s

## Checklist


## Expected behavior


- I can change tag format and still being able to:
	- bump versions (aka. find the increment and commits range)
	- generate changelog (full, incremental, version range)
- I can generate a changelog without warnings for expected tags

## Steps to Test This Pull Request
- create a project
- set `version_provider = "scm"` (not mandatory but allows testing the full scope)
- bump some versions with an initial `tag_format`
- move current `tag_format` in `legacy_tag_format` and define a new `tag_format`
- bump a version with changelog:
	- current version should be properly identified
	- increment should properly been found
	- new version should be correct
	- generated changelog generation should be correct (both new and old version formats)
	- new tag should respect the new `tag_format`

then:
- add `random-*` to `ignored_tag_formats`
- tag any commit with `random-tag`
- run cz changelog
	- there shouldn't be any warning

## Additional context

There might be a 2nd PR handling a `changelog_version_format` setting to customize the version rendering in the changelog so:
- it's not the tag format anymore (problematic with some prefixes, like for monorepos)
- have a uniform version display across `tag_format` changes
